### PR TITLE
Add print_layer_shapes function

### PR DESCRIPTION
**Pull request by julienr**
This adds a new utility function ``print_layer_shapes`` that can be used to print the output shape of each layer in a model. 

I am not sure this is the most efficient way to do it, but it seems to work.

For example :
```python
from keras.models import Sequential
from keras.layers.core import Layer, Dense, Dropout, Activation, Flatten, Reshape
from keras.layers.convolutional import Convolution2D, MaxPooling2D
from keras.utils.model_utils import print_layer_shapes

model = Sequential()

model.add(Convolution2D(32, 1, 3, 3, border_mode='valid')) 
model.add(Convolution2D(32, 32, 3, 3, border_mode='valid'))
model.add(MaxPooling2D(poolsize=(2, 2)))
model.add(Flatten())
model.add(Dense(32*12*12, 128))
model.add(Dense(128, 10))

print_layer_shapes(model, (1, 1, 28, 28))
```

Will print

```
input shape :  (1, 1, 28, 28)
shape after Convolution2D : [ 1 32 26 26]
shape after Convolution2D : [ 1 32 24 24]
shape after MaxPooling2D : [ 1 32 12 12]
shape after Flatten : [   1 4608]
shape after Dense : [  1 128]
shape after Dense : [ 1 10]
```

### CODE REVIEW

**Comment by iskandr** [comment]
It looks like this only works for `Sequential` models, any chance this could also be made to work with a `Graph`? 

Also, a significantly harder and more niche request: how about handling unknown input shapes by propagating symbolic names for those dimensions (e.g. make the input dims SymPy symbols with names like "x", "y", "z", "w" or "input0".."input3" and build up algebraic expressions for dim calculations that depend on them).

**Comment by fchollet** [comment]
Definitely useful for debugging.

Indeed, if we were to merge this it would need to be generic, i.e. work for any model (including Sequential models that include Merge layers, and Graph models).

**Comment by julienr** [comment]
@iskandr This is an interesting suggestion, but I think you usually know your input shape. For example in a CNN, even if the Convolution2D layer doesn't have to know the size of the images, the Dense layer that follows will have the image size specified. Do you have an example where having algebraic dimensions would be useful ?

**Comment by julienr** [comment]
@fchollet I added a separate function that works for Graph model. For Graph models, it would be nice to be able to print a computation tree for each output. But it seems there is currently no generic way to walk the graph. For example, for a given layer ``l``, its parent is at ``graph.nodes[l.previous.name]`` but : 
- Not all layers implement ``set_name`` (Merge doesn't implement it for example)
- ``Layer.set_name`` doesn't set ``self.name`` but just modifies the parameters.

Would it make sense to fix this by having all layers implement ``set_name`` and add a ``Layer.name`` attribute ?

**Comment by iskandr** [comment]
@julienr I was thinking this would be useful for debugging shape mismatch errors, especially those arising from incorrect assumptions about reshape/flatten/embedding. Instead of tracking constants through the computation graph, it might be easier to see that "wait, I didn't expect 'height * width' to come out of that node". 

It might, however, be sufficient just to see the all the constant shapes (bypassing the complexity of building up symbolic computations).

**Comment by pranv** [comment]
I had worked on this, as a part of my dimensionality inference code. I think it can be solved by having an optional attribute called `input_shape ` and `output_shape ` for each layer. When a new layer is added, it can get input shape to it by `get_input_shape ` method, which fetches the output shapes of all previous layers. And then we can set layer's `output_shape` after performing the calculations, which can be subsequently used further. This will create a recursive, on demand call, hence no symbolic stuff necessary. This can be used in graphs also, with a minor change.



I've the code to calculate the output shapes of almost all layers. But, I'm really busy, so I can share it with anyone willing to work on it.




Also, this is a much simpler way to help the user in this regard, without modifying the API significantly. So :+1:

**Comment by julienr** [comment]
@pranv I think what you propose is close to what [Lasagne does](http://lasagne.readthedocs.org/en/latest/modules/layers.html#lasagne.layers.Layer.get_output_shape) with its ``lasagne.layers.Layer.get_output_shape``. This is a great approach, the only downside is that this require duplicating the whole shape computation in keras. But it might be worth it as it makes it easier than my solution to get the shape of a specific layer.

**Comment by fchollet** [comment]
Merged. Had to rewrite it almost completely :( https://github.com/fchollet/keras/blob/master/keras/utils/layer_utils.py#L88

**Comment by julienr** [comment]
@fchollet Much better after rewrite :) Thanks
